src/SpeedLimiter.cc

```diff
@@ -97,105 +44,152 @@ class ignition::math::SpeedLimiterPrivate
+double SpeedLimiter::MinAcceleration() const
+{
+  return this->dataPtr->minAcceleration;
+}
+
+//////////////////////////////////////////////////
+void SpeedLimiter::SetMaxAcceleration(double _lim)
+{
+  this->dataPtr->maxAcceleration = _lim;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::MaxAcceleration() const
+{
+  return this->dataPtr->maxAcceleration;
+}
+
+//////////////////////////////////////////////////
+void SpeedLimiter::SetMinJerk(double _lim)
+{
+  this->dataPtr->minJerk = _lim;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::MinJerk() const
+{
+  return this->dataPtr->minJerk;
+}
+
+//////////////////////////////////////////////////
+void SpeedLimiter::SetMaxJerk(double _lim)
+{
+  this->dataPtr->maxJerk = _lim;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::MaxJerk() const
+{
+  return this->dataPtr->maxJerk;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::Limit(double &_vel, double _prevVel, double _prevPrevVel,
     std::chrono::steady_clock::duration _dt) const
 {
-  const double tmp = _v;
+  const double vDesired = _vel;
 
-  this->LimitJerk(_v, _v0, _v1, _dt);
-  this->LimitAcceleration(_v, _v0, _dt);
-  this->LimitVelocity(_v);
+  this->LimitJerk(_vel, _prevVel, _prevPrevVel, _dt);
+  this->LimitAcceleration(_vel, _prevVel, _dt);
+  this->LimitVelocity(_vel);
 
-  if (ignition::math::equal(tmp, 0.0))
-    return 1.0;
-  else
-    return _v / tmp;
+  return _vel - vDesired;
 }
 
 //////////////////////////////////////////////////
-double SpeedLimiter::LimitVelocity(double &_v) const
+double SpeedLimiter::LimitVelocity(double &_vel) const
 {
-  const double tmp = _v;
+  const double vDesired = _vel;
 
-  if (this->dataPtr->hasVelocityLimits)
-  {
-    _v = ignition::math::clamp(
-      _v, this->dataPtr->minVelocity, this->dataPtr->maxVelocity);
-  }
+  _vel = ignition::math::clamp(
+    _vel, this->dataPtr->minVelocity, this->dataPtr->maxVelocity);
 
-  if (ignition::math::equal(tmp, 0.0))
-    return 1.0;
-  else
-    return _v / tmp;
+  return _vel - vDesired;
 }
 
 //////////////////////////////////////////////////
-double SpeedLimiter::LimitAcceleration(double &_v, double _v0,
+double SpeedLimiter::LimitAcceleration(double &_vel, double _prevVel,
     std::chrono::steady_clock::duration _dt) const
 {
-  const double tmp = _v;
+  const double vDesired = _vel;
 
-  if (this->dataPtr->hasAccelerationLimits)
-  {
-    const double dvMin = this->dataPtr->minAcceleration *
-        std::chrono::duration<double>(_dt).count();
-    const double dvMax = this->dataPtr->maxAcceleration *
-        std::chrono::duration<double>(_dt).count();
+  const double dtSec = std::chrono::duration<double>(_dt).count();
 
-    const double dv = ignition::math::clamp(_v - _v0, dvMin, dvMax);
+  if (equal(dtSec, 0.0))
+    return 0.0;
 
-    _v = _v0 + dv;
-  }
+  const double accDesired = (_vel - _prevVel) / dtSec;
```

Comment:
I think the code is clearer to read this way, but I believe it may have been written the other way to avoid a division operation and divide by 0 concerns

Reply:
Yeah I agree that this may have been the motivation. One lesson I took from the fact that the math had been wrong on the `ros_control` implementation all this time is the value of readability. I'm leaning towards leaving the math straightforward so it's easier to debug in the future. I added tests for the `dt == 0` case, so I think we're covered.